### DISCUSSION TITLE
Dropdown - How to get choices to re-render when it's a callback?

### DISCUSSION THREAD

**Discussion by brancusi** [day8/re-com#274] [Q&A]
Hi, 

When using a callback for choices, and if something within that function should react to changes, the choices fn won't rerun leading to out of sync choices.

Example

```clojure
(defn uom-drop-down
  [{:keys [uoms
           model
           update-fn
           create-fn]}]
  [:div
   
   [:p (str "Total UOMS: " (count uoms))] ;; <- This updates fine

   [single-dropdown
    :src (at)
    :width "150px"
    :filter-box? true
    :auto-complete? true
    :model model
    :label-fn (fn [val] (:label val))
    :id-fn (fn [val] val)
    :debounce-delay 0
    :on-change (fn [val]
                 (case (:type val)
                   :create (create-fn (:val val))
                   :select (update-fn [(:id val) val])))
    :choices (fn [{:keys [filter-text]} done _] ;; <- This never gets recalled even if uoms changes
               (let [results (query->suggestions uoms :label filter-text)]
                 (done (if (seq results)
                         results
                         [{:label (str "No match found. Create " filter-text) :val filter-text :type :create}]))))]])
```

**_uoms_** is getting passed in from the parent. If there is no match, we want to be able to create a new UOM and then have that be added to the list for future use. Currently this choices won't rerun if **_uoms_** is updated.

Any help on how I might get this to work much appreciated.

Thank you

**Comment by superstructor** [comment]
This might be a state issue with how you've structured your reagent component `uom-drop-down`.

[Other reagent forms such as Form-2 or Form-3](https://github.com/reagent-project/reagent/blob/master/doc/CreatingReagentComponents.md) combined with [wrapping external state, see `external-model` and `latest-ext-model` in `input-text-base`](https://github.com/day8/re-com/blob/master/src/re_com/input_text.cljs#L65) might be helpful.
